diff -Naur kcoreaddons-5.49.0.orig/src/lib/util/kformat.h 
kcoreaddons-5.49.0/src/lib/util/kformat.h
--- kcoreaddons-5.49.0.orig/src/lib/util/kformat.h	2018-08-04 12:56:42.000000000 +0200
+++ kcoreaddons-5.49.0/src/lib/util/kformat.h	2018-08-14 12:35:57.224626936 +0200
@@ -116,7 +116,7 @@
      * @see formatValue
      * @since 5.49
      */
-    enum class Unit {
+    enum Unit {
         Other,
         Bit,   ///< "bit"
         Byte,  ///< "B"
@@ -135,7 +135,7 @@
      * @see formatValue
      * @since 5.49
      */
-    enum class UnitPrefix {
+    enum UnitPrefix {
         /// Auto-choose a unit such that the result is in the range [0, 1000 or 1024)
         AutoAdjust = -128,
