diff --git a/src/lib/CMakeLists.txt b/src/lib/CMakeLists.txt
index 573d6e3..8f0ff3a 100644
--- a/src/lib/CMakeLists.txt
+++ b/src/lib/CMakeLists.txt
@@ -205,7 +205,9 @@ ecm_generate_headers(KCoreAddons_HEADERS
     REQUIRED_HEADERS KCoreAddons_HEADERS
 )

-find_package(PythonModuleGeneration)
+if (GENERATE_PYTHON_MODULES)
+    find_package(PythonModuleGeneration)
+endif ()

 if (PythonModuleGeneration_FOUND)
   ecm_generate_python_binding(
