From eb47c08204deca0814fd84dcf53b45e611ac4442 Mon Sep 17 00:00:00 2001
From: Yuri Chornoivan <yurchor@ukr.net>
Date: Sun, 20 May 2018 08:14:50 +0300
Subject: Fix message extraction

---
 Messages.sh           | 2 +-
 scripting/Messages.sh | 2 --
 2 files changed, 1 insertion(+), 3 deletions(-)

diff --git a/Messages.sh b/Messages.sh
index bfe7801..477aa76 100644
--- a/Messages.sh
+++ b/Messages.sh
@@ -1,3 +1,3 @@
 #! /usr/bin/env bash
 $EXTRACTRC *.kcfg >> rc.cpp
-$XGETTEXT *.h *.cpp helpers/killer/*.cpp tabbox/*.cpp scripting/*.cpp -o $podir/kwin.pot
+$XGETTEXT *.h *.cpp helpers/killer/*.cpp plugins/scenes/opengl/*.cpp tabbox/*.cpp scripting/*.cpp -o $podir/kwin.pot
diff --git a/scripting/Messages.sh b/scripting/Messages.sh
index 022c0df..cd81a7f 100644
--- a/scripting/Messages.sh
+++ b/scripting/Messages.sh
@@ -1,4 +1,2 @@
 #! /usr/bin/env bash
-$EXTRACTRC `find . -name \*.rc -o -name \*.ui -o -name \*.kcfg` >> rc.cpp
 $XGETTEXT `find . -name \*.h -o -name \*.cpp` -o $podir/kwin_scripting.pot
-rm -f rc.cpp
-- 
cgit v0.11.2

